### DISCUSSION TITLE
Add raw signature to original PDF.

### DISCUSSION THREAD

**Discussion by Ort0x36** [MatthiasValvekens/pyHanko#322] [Q&A]
I have the raw signature of the SHA256 hash that I obtain when returning from the API, but I want to know how I can insert it into the PDF and validate it.

Example of request:

```
 {
     "hash": "Mjyw2pii29w6hg9XQca50xr3FzgJ/bmy5kpK1skMvOw="
 }
``` 

Example of response:

```
 {
     {"raw_signature": "-----BEGIN PKCS7-----\nMIAGCSqGSIb3DQEHAqCAMIACAQExDzANBglghkgBZQMEAgEFADCABgkqhkiG9w0BBwEAAKCAMIIHFjCCBP6gA\n-----END PKCS7-----"
 }
``` 

How can I add this raw signature to my original document? I was trying with [API PdfCMSEmbedder](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#pdf-cms-embedder-protocol) and [
in the section about interrupted subscription](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#interrupted-signing).

I'm saving my raw signature to a .p7s file. Here is the code I was testing:

```
from datetime import datetime
from pyhanko.sign import signers, fields
from pyhanko.sign.signers import cms_embedder
from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
import base64
from io import BytesIO

file = open('doc.pdf', 'rb')
input_buf = BytesIO(file.read())
w = IncrementalPdfFileWriter(input_buf)

cms_writer = cms_embedder.PdfCMSEmbedder().write_cms(
    field_name='Signature', writer=w
)
sig_field_ref = next(cms_writer)

assert sig_field_ref.get_object()['/T'] == 'Signature'


timestamp = datetime.now()
sig_obj = signers.SignatureObject(timestamp=timestamp, bytes_reserved=11916)

md_algorithm = 'sha256'

cms_writer.send(
    cms_embedder.SigObjSetup(
        sig_placeholder=sig_obj,
        mdp_setup=cms_embedder.SigMDPSetup(
            md_algorithm=md_algorithm, 
            certify=True,
            docmdp_perms=fields.MDPPerm.NO_CHANGES
        )
    )
)


prep_digest, output = cms_writer.send(
    cms_embedder.SigIOSetup(md_algorithm=md_algorithm, in_place=True)
)

hashb64 = base64.b64encode(prep_digest.document_digest).decode()

cms_bytes = open('documento.p7s', mode='rb').read()
sig_contents = cms_writer.send(cms_bytes)

with open("out_assinado.pdf", "wb") as outfile:
    outfile.write(output.getbuffer())    

```

**Comment by MatthiasValvekens** [comment]
Hi @Ort0x36,


That response string looks like a PEM-encoded PKCS#7/CMS message. So if you ["unarmor" it with `asn1crypto`](https://github.com/wbond/asn1crypto/blob/master/docs/pem.md#unarmor), you'll get the DER payload that `PdfCMSEmbedder` expects.

`asn1crypto` is a core dependency of pyHanko, so you should already have it installed.

**Reply by Ort0x36** [reply]
I did some tests with this and when loading the DER payload I am getting the following exception.

```
ValueError: Insufficient data - 118836279028905154188750128783647001770878995696700206383934836259634333809214610405699318030561097655377547780187443932071494980477516775373099009551197742151461011789 bytes requested but only 2491 available
``` 

Code:

```python
from asn1crypto import pem, x509

with open('./documento.p7s', 'rb') as f:
    der_bytes = f.read()
    if pem.detect(der_bytes):
        type_name, headers, der_bytes = pem.unarmor(der_bytes)

cert = x509.Certificate.load(der_bytes)
print(cert)

```

**Reply by MatthiasValvekens** [reply]
Ah, sorry, the code sample in the asn1crypto docs is for certificate parsing, whereas you're reading a CMS message. Either way, you don't need to do any parsing here. Just feed `der_bytes` to your CMS writer as-is, that should suffice.

(In case you're curious: with `cms.ContentInfo.load(...)` instead of `x509.Certificate.load(...)` it would've worked.)

**Reply by Ort0x36** [reply]
I understand, I believe that the way the code is right now, the signature should work, but I am validating the pdf in https://verificador.staging.iti.br/

But I still get invalid signature.
Reference example: [GOV BR signatures API integration roadmap](https://manual-integracao-assinatura-eletronica.servicos.gov.br/pt_BR/5.8/iniciarintegracao.html#exemplo-de-aplicacao)

Code:

```python
from datetime import datetime
from pyhanko.sign import signers, fields
from pyhanko.sign.signers import cms_embedder
from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
import base64
from io import BytesIO
from asn1crypto import pem


file = open('doc.pdf', 'rb')
input_buf = BytesIO(file.read())
w = IncrementalPdfFileWriter(input_buf)

cms_writer = cms_embedder.PdfCMSEmbedder().write_cms(
    field_name='Signature', writer=w
)
sig_field_ref = next(cms_writer)

assert sig_field_ref.get_object()['/T'] == 'Signature'

timestamp = datetime.now()
sig_obj = signers.SignatureObject(timestamp=timestamp, bytes_reserved=11916)

md_algorithm = 'sha256'

cms_writer.send(
    cms_embedder.SigObjSetup(
        sig_placeholder=sig_obj,
        mdp_setup=cms_embedder.SigMDPSetup(
            md_algorithm=md_algorithm, 
            certify=True,
            docmdp_perms=fields.MDPPerm.NO_CHANGES
        )
    )
)

prep_digest, output = cms_writer.send(
    cms_embedder.SigIOSetup(md_algorithm=md_algorithm, in_place=True)
)

hashb64 = base64.b64encode(prep_digest.document_digest).decode()


with open('./document.p7s', 'rb') as f:
    der_bytes = f.read()
    if pem.detect(der_bytes):
        type_name, headers, der_bytes = pem.unarmor(der_bytes)
        
sig_contents = cms_writer.send(der_bytes)

with open("out_sign.pdf", "wb") as outfile:
    outfile.write(output.getbuffer())    
```

**Reply by MatthiasValvekens** [reply]
If that's the code you're running as-is, then the issue is probably that you're trying to use the same API response across multiple calls ;)

Even if the underlying base document doesn't change, the document hash doesn't stay the same: the process of preparing a document for signing involves putting the current time somewhere into the doc, for example.

**Reply by Ort0x36** [reply]
I'm already adding the current time to the document at:

````python 
timestamp = datetime.now()
sig_obj = signers.SignatureObject(timestamp=timestamp, bytes_reserved=11916)
````

But I believe that the hash of the original document will always remain the same, as everything that is changed internally in the pdf document is saved in the output file that I am using to "validate" the signature.

**Reply by MatthiasValvekens** [reply]
Maybe so, but the fields in SignatureObject are part of the hashed content (except for the Contents entry itself). So the hash still changes 🙂 

PDF signatures are weird like that

**Reply by Ort0x36** [reply]
The biggest problem is that I receive a raw signature of the sha256 hash sent to the API, the worst part of all this is that the API documentation simply doesn't specify what I should do with this return, I've been working on this for a while and still haven't I achieved nothing.

**Reply by MatthiasValvekens** [reply]
AFAICT the issue is here:

```python
hashb64 = base64.b64encode(prep_digest.document_digest).decode()


with open('./document.p7s', 'rb') as f:
    der_bytes = f.read()
    if pem.detect(der_bytes):
        type_name, headers, der_bytes = pem.unarmor(der_bytes)
```

You really can't rely on `document_digest` being constant between runs. At this stage, you should really request a new signature from the service instead of trying to read the result of your previous fetch. Have you tried that?

**Reply by Ort0x36** [reply]
Explaining the code, the API expects the base64 hash of the document to perform the signature, which is why I use base64 encode in document_digest.

For each execution, I send the digital document in base64 for signature in the service and save the return in the document.p7s file. So theoretically for each execution a new hash is generated and signed.

When the API returns, I can receive the signed hash in RAW/PKCS1 or CMS/PKCS7 format, depending on the parameter I send.

I must be making some confusion with the return of this signed hash, as I can't put it inside the PDF.

My first question is, can I put any of these formats in the original PDF?

If yes, my second question, how do I load this signed hash in the right format to put in the PDF and transform it into a Signed PDF?

**Reply by Ort0x36** [reply]
Explaining the raw signature part better, assuming I use Pyhanko's internal "sign_raw" function, which returns the raw signature of a given hash, that is, this is practically what the API theoretically does internally.

```python
from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
from pyhanko.sign import signers

signer = signers.SimpleSigner.sign_raw(data=b'', digest_algorithm='') # # get raw_signature
```

What do I need to embed this raw signature inside my PDF?

**Reply by MatthiasValvekens** [reply]
(on mobile, sorry for any typos/curtness)

Actually it's not; you get back a full PKCS#7/CMS payload. `sign_raw` would imply only returning the serialised signature of the signed attributes (there are remote signers that do that, but from your previous comments yours seems to be different).

I think you're getting overly hung up on "embedding the signature into the original PDF". That's just not how PDF signing works: the process of preparing the document for signing _cannot_ be decoupled from the document hash computation.

You can't "prefetch" the signature and then expect to be able to embed it into the original file from scratch. PyHanko's interrupted signing API supplies a "prepared" PDF data stream together with the document hash. You have to use that hash to retrieve the signature (that's the vendor-dependent part), and you have to insert it into the prepared stream that you got together with that hash.

**Reply by Ort0x36** [reply]
In that case, what should I do then? my situation is extremely similar to the discussion: #90

I also have a vendor API endpoint to retrieve the user's certificate that returns a PEM string, I don't know if this works for me, but really this is all a big headache. What would you recommend me?

**Reply by MatthiasValvekens** [reply]
Have you taken a look at [this example](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#interrupted-signing-when-the-signer-s-certificate-is-not-known-a-priori) in the documentation? Essentially, the step that you'd have to implement is `call_external_service` to obtain the signature from your external service given the hash pyHanko computes for you. Assuming the response is of the form you specified in your first post, e.g.

```
 {
     {"raw_signature": "-----BEGIN PKCS7-----\nMIAGCSqGSIb3DQEHAqCAMIACAQExDzANBglghkgBZQMEAgEFADCABgkqhkiG9w0BBwEAAKCAMIIHFjCCBP6gA\n-----END PKCS7-----"
 }
```

the only transformation you have to apply is extracting the DER from this `raw_signature` value (using `pem.unarmor` from `asn1crypto`, as we discussed above). The value resulting from that is the `signature_container` in the code sample.

I suggest doing your first implementation without trying to optimise resource usage--just fetch the signature from the service while keeping all pyHanko objects in memory, don't try to save anything other than the final output to disk. Once you get that working, you can start thinking about how to persist the document's state while waiting for the response.
